### ISSUE TITLE
Google Chrome back bug from image

### ISSUE THREAD

**Issue by bobintetley** [bobintetley/asm3#716]
When clicking on the thumbnail on the animal details record in Chrome 70, it takes you to the image. If you go back, the field contents have all moved by one slot (all sliders). If you save the record, it saves with the values in the wrong fields.

As a simple fix, have the thumbnail open the image in a new tab to prevent this.

**Comment by bobintetley** [comment]
Eg: code is now in name, microchip date is in microchip number, etc.

**Comment by jgardynik** [comment]
Since I'm using the self-hosted version that's still back on 42 released almost 9 months ago (where your bugfixes are not getting propagated to), and because this fix is merely glossing over the actual problem dealing specifically with clicking on photos, I figured I'd let you know the real fix for this. Put `autocomplete="off"` on all the input fields. Chrome's not likely going to fix this, since it's been around many many years now (Since version 28 at least), and is the result of a strange interaction with jQuery, Chrome's state saving mechanism, and accessing elements by ID only. And the above is the routinely thrown around fix on stackoverflow and jQuery's forums. I'm honestly surprised you haven't run into it before now.

As an aside, are you planning on moving to jQuery 3.4 anytime soon? JQuery2's been deprecated for years now, and the version included in ASM is almost 5 years old now. Given how heavily this relies on AJAX requests, I wouldn't be surprised if it were vulnerable to some of the XSS attacks that older version of JQuery have.

I'll also ask again if we could get a release sometime soon. 9 months is a long time to be sitting on things.

**Comment by bobintetley** [comment]
I've run into autocomplete problems with Chrome previously and we are setting the autocomplete attribute on some inputs already (mainly dates). 

I've verified that the state saving bug was not present before Chrome 70 and I don't really see how autocomplete is related to it, but I'll do some tests to see if it helps.

Yes, JQuery and UI/Mobile dependencies will be upgraded soon.

There's still some tickets left that we'd like to be in 43 before it's released. 43 adds support for Python 3 and we had been waiting on a fix to python3-webpy in Debian (they've retired python-webpy from the buster repository).

The trunk is stable and can be run.

**Comment by bobintetley** [comment]
Spammed autocomplete="off" to every input and tested with Chromium 79. The bug still occurs. I tried other variants, like autocomplete="nope" and autocomplete="new-password". No effect, it doesn't look like autocomplete is related. Other browsers do not have this bug and neither does Chrome prior to 70.

**Comment by jgardynik** [comment]
That's weird, because I did just the same, and it fixed every field they tried to move into in Chrome 80. Before, If I clicked on the image and then went back, it would shift every element down one. If I did it again, it would shift it _again_ by another one element. This resulted in dates and prices in our additional fields section. After setting `autocomplete='off'` on all of our additional fields it still would happen, but they'd move even farther down to health info. I guess even if you add it to them all, it'd probably still move them one (although where they'd go I couldn't tell you). Can't really test setting the entire form to off because there's no form element.

I wonder if the real solution is to run a function to check some form of idempotency token and if it's not right then force clear the entire form and re-apply all the values. Or just `location.reload()` if you can tell that it's stale data. I can look into it a little further, although it's a bit of a pain to debug with all the deferred and promise'd function calls.

Do you have a relevant bug report or problem that is the actual cause from Chrome 70? I can also look through the versions to bisect what commit broke things for good.

**Comment by bobintetley** [comment]
Mine rotated by 1 whether the autocomplete attribute was present or not. 

I agree, forcing a reload when detecting that the page been visited via back might be the best and simplest way to deal with the problem.

**Comment by bobintetley** [comment]
I tested out a fix that works by detecting that the page was navigated to via the back button and it seems to work for me. Patch references this bug. Thanks for the recommendation.

**Comment by jgardynik** [comment]
The good news is, I think I found your culprit, too.

https://chromium.googlesource.com/chromium/src/+/8bdd4fc873801be72f20f7cb5746059526098d99

Chromium version 79.0.3925. It works before this, and not afterward. Given that the whole commit is dealing with saved control states, it's pretty surefire. Unfortunately, it looks like the whole point of this is to fix another problem, so the idea of just force-refreshing on back is probably still easier. Your other option might be to implement a slight delay before trying to assign your values. Let Chrome do whatever it's trying to do, then fill out the values. 

You can test the following versions to see it break (if you're on windows):
https://storage.googleapis.com/chromium-browser-snapshots/index.html?prefix=Win_x64/700567/
https://storage.googleapis.com/chromium-browser-snapshots/index.html?prefix=Win_x64/700575/

**Comment by bobintetley** [comment]
Very interesting. Thank you for the research on this. That explains a lot if it's a race condition. There are more areas that need the fix and ideally it would be better to put the fix into the module loading workflow instead to avoid having to repeat the code on multiple screens.
